src/main/java/org/nixos/idea/lang/NixFoldingBuilder.kt

```diff
@@ -0,0 +1,135 @@
+
+        root.accept(object : NixElementVisitor<Unit>() {
+            override fun visitExprAttrs(o: NixExprAttrs) {
+                o.acceptChildren(this)
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+                o.bindList.forEach { it.accept(this) }
+            }
+
+            override fun visitBindAttr(o: NixBindAttr) {
+                o.expr?.accept(this)
+            }
+
+            override fun visitExprParens(o: NixExprParens) {
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+                o.expr?.accept(this)
+            }
+
+            override fun visitExprWith(o: NixExprWith) {
+                o.exprList.forEach { it.accept(this) }
+            }
+
+            override fun visitExprApp(o: NixExprApp) {
+                o.exprList.forEach { it.accept(this) }
+            }
+
+            override fun visitExprIf(o: NixExprIf) {
+                o.acceptChildren(this)
+            }
+
+            override fun visitExprAssert(o: NixExprAssert) {
+                o.exprList.forEach { it.accept(this) }
+            }
+
+            override fun visitIndString(o: NixIndString) {
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+            }
+
+            override fun visitExprOpUpdate(o: NixExprOpUpdate) {
+                o.acceptChildren(this)
+            }
+
+            override fun visitExprOpConcat(o: NixExprOpConcat) {
+                o.acceptChildren(this)
+            }
+
+            override fun visitExprList(o: NixExprList) {
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+                o.items.forEach { it.accept(this) }
+            }
+
+            override fun visitExprSelect(o: NixExprSelect) {
+                o.acceptChildren(this)
+            }
+
+            override fun visitExprLet(o: NixExprLet) {
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+                o.bindList.forEach { it.accept(this) }
+                o.expr?.accept(this)
+            }
+
+            override fun visitFormals(o: NixFormals) {
+                descriptors.add(FoldingDescriptor(o, o.textRange))
+                o.formalList.forEach { it.accept(this) }
+            }
+
+            override fun visitComment(comment: PsiComment) {
+                descriptors.add(FoldingDescriptor(comment, comment.textRange))
+            }
+
+            override fun visitFormal(o: NixFormal) {
+                o.expr?.accept(this)
+            }
+
+            override fun visitExprLambda(o: NixExprLambda) {
+                o.expr?.accept(this)
+                o.formals?.accept(this)
+            }
+
+            override fun visitFile(file: PsiFile) {
+                super.visitFile(file)
+                file.acceptChildren(this)
+            }
+        })
+
+        return descriptors.toArray(FoldingDescriptor.EMPTY_ARRAY)
+    }
+
+    override fun getPlaceholderText(node: ASTNode): String? =
```

Comment:
thought: Looks like you could provide the placeholder text and whether the region is collapsed by default to the constructor of `FoldingDescriptor`. It looks like in this case, `getPlaceholderText` and `isCollapsedByDefault` could just throw an `UnsupportedOperationException` or trigger an assertion?